packages/core-data/src/resolvers.js

```diff
@@ -273,34 +273,42 @@ export const getEmbedPreview =
  * Checks whether the current user can perform the given action on the given
  * REST resource.
  *
- * @param {string}  action   Action to check. One of: 'create', 'read', 'update',
- *                           'delete'.
- * @param {string}  resource REST resource to check, e.g. 'media' or 'posts'.
- * @param {?string} id       ID of the rest resource to check.
+ * @param {string}  requestedAction Action to check. One of: 'create', 'read', 'update',
+ *                                  'delete'.
+ * @param {string}  resource        REST resource to check, e.g. 'media' or 'posts'.
+ * @param {?string} id              ID of the rest resource to check.
  */
 export const canUser =
-	( action, resource, id ) =>
-	async ( { dispatch } ) => {
-		const methods = {
-			create: 'POST',
-			read: 'GET',
-			update: 'PUT',
-			delete: 'DELETE',
-		};
+	( requestedAction, resource, id ) =>
+	async ( { dispatch, registry } ) => {
+		const { hasStartedResolution } = registry.select( STORE_NAME );
+
+		const resourcePath = id ? `${ resource }/${ id }` : resource;
+		const retrievedActions = [ 'create', 'read', 'update', 'delete' ];
 
-		const method = methods[ action ];
-		if ( ! method ) {
-			throw new Error( `'${ action }' is not a valid action.` );
+		if ( ! retrievedActions.includes( requestedAction ) ) {
+			throw new Error( `'${ requestedAction }' is not a valid action.` );
 		}
 
-		const path = id
-			? `/wp/v2/${ resource }/${ id }`
-			: `/wp/v2/${ resource }`;
+		// Prevent resolving the same resource twice.
+		for ( const relatedAction of retrievedActions ) {
+			if ( relatedAction === requestedAction ) {
+				continue;
+			}
+			const isAlreadyResolving = hasStartedResolution( 'canUser', [
+				relatedAction,
+				resource,
+				id,
+			] );
+			if ( isAlreadyResolving ) {
+				return;
+			}
+		}
```

Comment:
Shouldn't this be built into resolvers, instead of modifying a specific resolver? Or is that not possible?

Reply:
Ah, got it, you're matching similar resolvers.